src/server/endpoints/covidcast_utils/model.py

```diff
@@ -299,3 +311,263 @@ def map_row(source: str, signal: str) -> str:
+
+    Example:
+    [TimePair("day", [20210407])] with pad_length 6 would return [TimePair("day", [20210407]), TimePair("day", [(20210401, 20210407)])].
+
+    Used to pad the user DB query with extra days.
+    """
+    if pad_length < 0:
+        raise ValueError("pad_length should non-negative.")
+    if pad_length == 0:
+        return time_pairs.copy()
+    min_time = min(time_value if isinstance(time_value, int) else time_value[0] for time_pair in time_pairs if not isinstance(time_pair.time_values, bool) for time_value in time_pair.time_values)
+    padded_time = TimePair("day", [(shift_time_value(min_time, -1 * pad_length), min_time)])
+    new_time_pairs = time_pairs.copy()
+    new_time_pairs.append(padded_time)
+    return new_time_pairs
+
+
+def pad_time_window(time_window: Tuple[int, int], pad_length: int) -> Tuple[int, int]:
+    """Extend a time window on the left by pad_length.
+
+    Example:
+    (20210407, 20210413) with pad_length 6 would return (20210401, 20210413).
+
+    Used to pad the user DB query with extra days.
+    """
+    if pad_length < 0:
+        raise ValueError("pad_length should non-negative.")
+    if pad_length == 0:
+        return time_window
+    min_time, max_time = time_window
+    return (shift_time_value(min_time, -1 * pad_length), max_time)
+
+
+def get_day_range(time_pairs: Union[TimePair, List[TimePair]]) -> TimePair:
+    """Combine a list of TimePairs into a single contiguous, explicit TimePair object.
+
+    Example:
+    [TimePair("day", [20210407, 20210408]), TimePair("day", [(20210408, 20210410)])] would return
+    TimePair("day", [20210407, 20210408, 20210409, 20210410]).
+
+    Used to produce a contiguous time index for time series operations.
+    """
+    if isinstance(time_pairs, TimePair):
+        time_pair = time_pairs
+        time_values = sorted(list(set().union(*(set(time_value_range(time_value)) if isinstance(time_value, tuple) else {time_value} for time_value in time_pair.time_values))))
+        if True in time_values:
+            raise ValueError("TimePair.time_value should not be a bool when calling get_day_range.")
+        return TimePair(time_pair.time_type, time_values)
+    elif isinstance(time_pairs, list):
+        if not all(time_pair.time_type == "day" for time_pair in time_pairs):
+            raise ValueError("get_day_range only supports day time_type pairs.")
+        time_values = sorted(list(set().union(*(get_day_range(time_pair).time_values for time_pair in time_pairs))))
+        return TimePair("day", time_values)
+    else:
+        raise ValueError("get_day_range received an unsupported type as input.")
+
+
+def _generate_transformed_rows(
+    parsed_rows: Iterable[Dict], time_pairs: Optional[List[TimePair]] = None, transform_dict: Optional[Dict[Tuple[str, str], List[Tuple[str, str]]]] = None, transform_args: Optional[Dict] = None, group_keyfunc: Optional[Callable] = None, data_signals_by_key: Dict[Tuple[str, str], DataSignal] = data_signals_by_key,
```

Comment:
So the whole relationship between generator, generator iterator, and iterator feels annoyingly murky to me, but I think I'm settling down on this interpretation of the docs:
- a [generator](https://docs.python.org/3/glossary.html#term-generator) is a function with a yield statement; when instantiated it returns a generator iterator
- a [generator iterator](https://docs.python.org/3/glossary.html#term-generator-iterator) is an iterator made by a generator function
- an [iterator](https://docs.python.org/3/glossary.html#term-iterator) is an object with a `__next__` method defined

So since `parsed_rows` is not a function that makes an iterator, but instead is the result of the [generator expression](https://docs.python.org/3/glossary.html#term-generator-expression) `parsed_rows = (parse_row(row, fields_string, fields_int, fields_float) for row in rows)` in `covidcast.py`, I'm going to mark this as `Iterator` and call it a day. Technically this function should work regardless of where parsed_rows came from, as long as it is an iterator that returns dicts (so it could work with a list or an iterator or a generator iterator).

I will mention in the docs though that we're assuming that `parsed_rows` is an iterable over rows from a query in the database.